### ISSUE TITLE
LOAM Based Localization Task 4 - Implement feature matching

### ISSUE THREAD

**Issue by ataparlar** [autowarefoundation/autoware.universe#7452]
### Checklist

- [X] I've read the [contribution guidelines](https://github.com/autowarefoundation/autoware/blob/main/CONTRIBUTING.md).
- [X] I've searched other issues and no duplicate issues were found.
- [X] I've agreed with the maintainers that I can plan this task.

### Description

In order to find the pose of the ego vehicle in the environment, the transform between each point cloud frame and the edge feature map must be known. 

### Purpose

Completing the feature matching between point cloud frame and previously generated edge point cloud.

### Possible approaches

- This repository can be used as a reference: [https://github.com/tier4/lidar_feature_extraction/tree/develop](https://github.com/tier4/lidar_feature_extraction/tree/develop)

### Definition of done

- [x] Match the point cloud features
- [ ] Find the transform between them

**Comment by ataparlar** [comment]
Hi everyone,

I am working with LIO-SAM directly because it is our implementation source. 

When I change the feature matching point cloud source with our ground truth feature maps instead of the map generated by itself and configure the program the following results occur.

The turquoise points are the corner feature cloud that we generated before. In the first two photos, we observed that it matches the features well and puts the `lidar_link`into the correct position.
In the 3rd and 4th images, while turning in the tunnel, the map corrupts. The position comes from IMU processing is got too far from the LiDAR feature matched position. This could be related to IMU data usage in the LIO-SAM.
![Screenshot from 2024-07-26 23-43-19](https://github.com/user-attachments/assets/afc55ba5-eefb-4d09-8e30-0fb34200d1e4)
![Screenshot from 2024-07-26 23-43-55](https://github.com/user-attachments/assets/f75515d2-c431-49ec-9781-a825b25e1add)
![Screenshot from 2024-07-26 23-45-54](https://github.com/user-attachments/assets/e6fd1806-a171-4c59-a0c5-6dff123992dc)
![Screenshot from 2024-07-26 23-46-15](https://github.com/user-attachments/assets/cf7e043b-329e-47dd-96d3-bcc2cd56ac65)


The below images are taken in a different test with a feature point cloud map that relies on over 20 kilometers. LIO-SAM could not process any correct `lidar_link` position in this test and didn't move in to correct direction. 
![Screenshot from 2024-07-27 01-41-02](https://github.com/user-attachments/assets/c0c15307-c53c-4ca4-a79d-9b6e3bd634a9)

Depending on the situation, I can try it with dynamic point cloud loading or directly try with LOAM instead of LIO-SAM. More updates will follow.

**Comment by ataparlar** [comment]
I tried to downsample the point clouds. It really improved the performance. It works well with tunnel and bridge datasets. Rosbags were running at low frequency. The only way to observe it was to slow down the rosbags because of the computational cost of larger point cloud maps. Here are some images:

![Screenshot from 2024-07-29 22-30-50](https://github.com/user-attachments/assets/d245c2ea-3062-4207-aa96-9141c5e6d1fd)
![Screenshot from 2024-07-29 22-28-14](https://github.com/user-attachments/assets/8f3ab2ec-aa3d-4881-9acb-6a7c877a1a58)

We may use this with dynamic point cloud loading for Autoware and then it will work like a charm in my opinion.

**Comment by ataparlar** [comment]
I have closed the GPS integration and loop closure detection. So, the code works faster with denser point clouds. 

I noticed an error in the pipeline. It can be related to point cloud undistortion or calibration angles. I changed the calibration angles and I observed no difference.

Here are the images and videos of the error.

[https://youtu.be/8XVvodyqYUI](https://youtu.be/8XVvodyqYUI)

White points show the ground truth point cloud to match the features.
Colored points are the points overlayed on the exported TF.
 
![Screenshot from 2024-08-14 21-43-03](https://github.com/user-attachments/assets/09a3eb64-12d1-43ac-a76e-e4fb5be25851)

![Screenshot from 2024-08-14 21-42-41](https://github.com/user-attachments/assets/a1a92a13-656f-4726-929a-37dc12132d93)

**Comment by ataparlar** [comment]
That is caused by an error in point cloud undistortion. The correction with the linear acceleration was disabled. When it is enabled, the problem eliminated.

**Comment by ataparlar** [comment]
The optimization part lasts long right now. The total time of scan to map optimization for each frame is more than 650 milliseconds. 
Here is the debug output of `loam_feature_localization`. Please note that the time numbers are in **milliseconds**:

```
[component_container_mt-1] end_update_initial_guess time: 0
[component_container_mt-1] extract_surrounding_key_frames time: 0
[component_container_mt-1] downsample_current_scan time: 0
[component_container_mt-1] 	corner_optimization time: 2
[component_container_mt-1] 	surface_optimization time: 14
[component_container_mt-1] 	combine_optimization_coeffs time: 0
[component_container_mt-1] 	corner_optimization time: 2
[component_container_mt-1] 	surface_optimization time: 14
[component_container_mt-1] 	combine_optimization_coeffs time: 0
[component_container_mt-1] 	corner_optimization time: 2
[component_container_mt-1] 	surface_optimization time: 14
[component_container_mt-1] 	combine_optimization_coeffs time: 0
[component_container_mt-1] 	corner_optimization time: 2
[component_container_mt-1] 	surface_optimization time: 15
[component_container_mt-1] 	combine_optimization_coeffs time: 0
[component_container_mt-1] 	corner_optimization time: 2
[component_container_mt-1] 	surface_optimization time: 15
[component_container_mt-1] 	combine_optimization_coeffs time: 0
[component_container_mt-1] 	corner_optimization time: 2
[component_container_mt-1] 	surface_optimization time: 15
[component_container_mt-1] 	combine_optimization_coeffs time: 0
[component_container_mt-1] 	corner_optimization time: 2
[component_container_mt-1] 	surface_optimization time: 15
[component_container_mt-1] 	combine_optimization_coeffs time: 0
[component_container_mt-1] 	corner_optimization time: 2
[component_container_mt-1] 	surface_optimization time: 15
[component_container_mt-1] 	combine_optimization_coeffs time: 0
[component_container_mt-1] 	corner_optimization time: 2
[component_container_mt-1] 	surface_optimization time: 15
[component_container_mt-1] 	combine_optimization_coeffs time: 0
[component_container_mt-1] 	corner_optimization time: 2
[component_container_mt-1] 	surface_optimization time: 15
[component_container_mt-1] 	combine_optimization_coeffs time: 0
[component_container_mt-1] 	corner_optimization time: 2
[component_container_mt-1] 	surface_optimization time: 15
[component_container_mt-1] 	combine_optimization_coeffs time: 0
[component_container_mt-1] 	corner_optimization time: 2
[component_container_mt-1] 	surface_optimization time: 15
[component_container_mt-1] 	combine_optimization_coeffs time: 0
[component_container_mt-1] 	corner_optimization time: 2
[component_container_mt-1] 	surface_optimization time: 15
[component_container_mt-1] 	combine_optimization_coeffs time: 0
[component_container_mt-1] 	corner_optimization time: 2
[component_container_mt-1] 	surface_optimization time: 15
[component_container_mt-1] 	combine_optimization_coeffs time: 0
[component_container_mt-1] 	corner_optimization time: 2
[component_container_mt-1] 	surface_optimization time: 15
[component_container_mt-1] 	combine_optimization_coeffs time: 0
[component_container_mt-1] 	corner_optimization time: 2
[component_container_mt-1] 	surface_optimization time: 15
[component_container_mt-1] 	combine_optimization_coeffs time: 0
[component_container_mt-1] 	corner_optimization time: 2
[component_container_mt-1] 	surface_optimization time: 15
[component_container_mt-1] 	combine_optimization_coeffs time: 0
[component_container_mt-1] 	corner_optimization time: 2
[component_container_mt-1] 	surface_optimization time: 15
[component_container_mt-1] 	combine_optimization_coeffs time: 0
[component_container_mt-1] 	corner_optimization time: 2
[component_container_mt-1] 	surface_optimization time: 15
[component_container_mt-1] 	combine_optimization_coeffs time: 0
[component_container_mt-1] 	corner_optimization time: 2
[component_container_mt-1] 	surface_optimization time: 15
[component_container_mt-1] 	combine_optimization_coeffs time: 0
[component_container_mt-1] 	corner_optimization time: 2
[component_container_mt-1] 	surface_optimization time: 15
[component_container_mt-1] 	combine_optimization_coeffs time: 0
[component_container_mt-1] 	corner_optimization time: 2
[component_container_mt-1] 	surface_optimization time: 15
[component_container_mt-1] 	combine_optimization_coeffs time: 0
[component_container_mt-1] 	corner_optimization time: 2
[component_container_mt-1] 	surface_optimization time: 15
[component_container_mt-1] 	combine_optimization_coeffs time: 0
[component_container_mt-1] 	corner_optimization time: 2
[component_container_mt-1] 	surface_optimization time: 15
[component_container_mt-1] 	combine_optimization_coeffs time: 0
[component_container_mt-1] 	corner_optimization time: 2
[component_container_mt-1] 	surface_optimization time: 15
[component_container_mt-1] 	combine_optimization_coeffs time: 0
[component_container_mt-1] 	corner_optimization time: 2
[component_container_mt-1] 	surface_optimization time: 15
[component_container_mt-1] 	combine_optimization_coeffs time: 0
[component_container_mt-1] 	corner_optimization time: 2
[component_container_mt-1] 	surface_optimization time: 15
[component_container_mt-1] 	combine_optimization_coeffs time: 0
[component_container_mt-1] 	corner_optimization time: 2
[component_container_mt-1] 	surface_optimization time: 15
[component_container_mt-1] 	combine_optimization_coeffs time: 0
[component_container_mt-1] 	corner_optimization time: 2
[component_container_mt-1] 	surface_optimization time: 15
[component_container_mt-1] 	combine_optimization_coeffs time: 0
[component_container_mt-1] 	corner_optimization time: 2
[component_container_mt-1] 	surface_optimization time: 15
[component_container_mt-1] 	combine_optimization_coeffs time: 0
[component_container_mt-1] scan_to_map_optimization time: 664
```

**Comment by ataparlar** [comment]
We noticed some problems related to reading the points. Normally, each point in the dataset has a 16-bit unsigned integer time field inside it. The program tries to read it as float. So, the time difference between the points cannot be detected and the point cloud undistortion never worked. We have fixed that.
Additionally, we noticed something with acceleration directions in the dataset and fixed that in the code. Now, the program works fluently without breaking. Here is the situation:

- Red points: Surface Map Points
- Turquoise points: Corner Map Points
- Yellow Points: Surface Cloud Points
- Blue Points: Corner Cloud Points
![Screenshot from 2024-09-24 09-46-21](https://github.com/user-attachments/assets/74b91d15-2626-49e2-aac0-70a2cdacbdaf)


In the video below, you can see the performance in the Eurasia Tunnel (the long tunnel). In the end, LOAM stuck but we solved that with IMU bias parameters. So, we can get in and get out of the Eurasia Tunnel.
[https://www.youtube.com/watch?v=lo1u--8MLwQ](https://www.youtube.com/watch?v=lo1u--8MLwQ)

We recorded a different video after the tunnel. We travel here on a highway and travel the Bosphorus Bridge on this part. After the city environment, we get into the Bomonti Tunnel (small tunnel) and the feature matching fails here:
[https://www.youtube.com/watch?v=JO3q8bNIxqs](https://www.youtube.com/watch?v=JO3q8bNIxqs) This video only contains the area after the Eurasia Tunnel. As you can see, the rosbag leaps frames. This is related to my ethernet settings at that moment. So, the error in the Bomonti Tunnel might be related to that because we have successfully traversed this part in a different dataset.

It doesn't break in the long tunnel, because there are very many features to match. In the small tunnel, there are fewer and sometimes it struggles to find the perfect position. I am working on it and I will keep here posted. It is nearly done.

**Comment by ataparlar** [comment]
Hi, 

I have worked on a modified LIO-SAM repo until now. [Here is the link](https://github.com/ataparlar/LIO-SAM/tree/ros2). This branch works with MGRS point cloud map and the XYZ coordinates of all of the points were very large. So, we are translating the point cloud segments to local and calculating the position here. After that, we translate back the calculated pose to the global. In that scenario, the pose is calculated fine but due to the large coordinate numbers, imuPreintegration can corrupt or precision reduces for position calculation. In the end, the position inside of a tunnel can't be calculated. It is stuck. 

[In this branch](https://github.com/ataparlar/LIO-SAM/tree/local-map), which works with a local point cloud map, we do not have that many large coordinates. So, we don't translate anything to local and calculate everything as where it is. In that scenario, we are able to travel the long tunnel (this one has more features) and we are able to travel every urban or highway area. However, it is stuck in the second tunnel. It is 3-4 kilometers far from the origin and it has much fewer features than the long tunnel. Still, we were able to travel that tunnel with a map starting from the start of the tunnel. We have tested that with previously collected data before.

So, as far as I observed, the distance from the origin is the issue right now. Used libraries don't accept `double` type so I couldn't change it.
